doc/ni_by_example_3.md

```diff
@@ -1,243 +1,356 @@
+1
+1
+1
+1
+1
+2
 (END)
 ```
 
-`ni` also works with the bash expansion operator `*`.
+Clearly this regex is operating on each line (the lines input were the integers 1 through 20). The way this works is another piece of Perl's uncompromising commitment to coding efficiency; default variables.
 
-For example, `ni dir_test/*` yields:
+In fact, the code above is operating on the most important of the Perl default variables, `$_`, which stores the value of the input line. Note that it shares the same name as the variable `@_`, and the similar-looking `$_[...]` will reference one of the scalars in `@_`, and not one of the characters in `$_`. Python-style string slicing (which uses square brackets) is not one available in vanilla Perl 5, where the `substr` method (or regexes) are more commonly used.
+
+
+##Intermediate Perl Operations
+####`p'^{...} ...'`: Begin Block
+A begin block is indicated by attaching a caret (`^`) to a block of code (encolsed in `{ }`). Outside of begin blocks, the Perl code is evaluated for every row; inside a Begin Block, the code is evaluated once and factored over the entire remaining Perl code. The example from the previous section:
 
 ```
-hello
-you
+$ ni n1p'sub yo {"hi " . $_[0]} yo a'
+hi 1
 (END)
 ```
 
-`ni` is able to go to the files directly becasue it is applies bash expansion first; bash expansion generates the file paths, which `ni` is then able to interpret.
+Would be better written as:
 
 ```
-$ ni --explain dir_test/*
-["cat","dir_test/hi"]
-["cat","dir_test/there"]
+$ ni n1p'^{sub yo {"hi " . $_[0]}} yo a'
+hi 1
+(END)
 ```
 
-##Intermediate Column Operations
-We can weave together row, column, and Perl operations to create more complex row operations. We also introduce some more advanced column operators.
+Begin Blocks are also useful for initializing data structures that will be manipulated, and in particular for converting data closures to Perl data structures, as we will see later in this section.
 
-* `w`: Append column to stream
-  * `$ ni <data> w[np'a*a']`
-  * `w` will add columns only up to the length of the input stream
-* `W`: Prepend column stream
-  * `$ ni <data> Wn` - Add line numbers to the stream (by prepending one element the infinite stream `n`)
-  * `W` will add rows only up to the length of the input stream
-* `v`: Vertical operation on columns
-  * **Important Note**: As of 2016-12-23, this operator is too slow to use in production.
+#### `a_` through `l_`: Multiline Selection operations 
+You have seen one way to generate multiple lines through the production of data closures. In order to access data from a specific column of the data closure, you will need to use multiline operators `a_` through `l_`, which are the multiline analogs to the line-based operators `a/a()` through `l/l()`.
 
+For example:
 
+* `$ ni ::data[n5] n1p'a(data)'` and `$ ni ::data[n5] n1p'a data'` will raise syntax errors, since `a/a()` are not prepared to deal with the more than one line data in the closure.
```

Comment:
Nitpick: technically it's a runtime prototype-mismatch error that happens here. Although prototypes influence the parser, Perl doesn't actually verify the argument count until invocation-time.